### PR TITLE
Fixes #5050 and #11826

### LINKED ISSUE #5050: push pragma is silently ignored if position is in front of import statement
That can be very confusing.

(And I would really prefer SmallLshouldNotBeUsed being off by default :-)

```
cat t.nim
{.push warning[SmallLshouldNotBeUsed]: off.}

import strutils

var l = 1
echo l
stefan@nuc ~/NEd $ nimsuggest --v2 --stdin t.nim
Hint: used config file '/home/stefan/Nim/config/nim.cfg' [Conf]
Hint: used config file '/home/stefan/NEd/nim.cfg' [Conf]
usage: sug|con|def|use|dus|chk|highlight|outline file.nim[;dirtyfile.nim]:line:col
type 'quit' to quit
type 'debug' to toggle debug mode on/off
type 'terse' to toggle terse mode on/off
> chk t.nim:1:0
Hint: t [Processing]
t.nim(5, 5) Warning: 'l' should not be used as an identifier; may look like '1' (one) [SmallLshouldNotBeUsed]
t.nim(6, 6) Warning: 'l' should not be used as an identifier; may look like '1' (one) [SmallLshouldNotBeUsed]

> quit
stefan@nuc ~/NEd $ nim c t.nim 
Hint: used config file '/home/stefan/Nim/config/nim.cfg' [Conf]
Hint: used config file '/home/stefan/NEd/nim.cfg' [Conf]
Hint: system [Processing]
Hint: t [Processing]
Hint: strutils [Processing]
Hint: parseutils [Processing]
Hint: math [Processing]
Hint: algorithm [Processing]
t.nim(5, 5) Warning: 'l' should not be used as an identifier; may look like '1' (one) [SmallLshouldNotBeUsed]
t.nim(6, 6) Warning: 'l' should not be used as an identifier; may look like '1' (one) [SmallLshouldNotBeUsed]
CC: t
CC: stdlib_strutils
CC: stdlib_parseutils
CC: stdlib_math
CC: stdlib_algorithm
Hint:  [Link]
Hint: operation successful (14346 lines compiled; 0.178 sec total; 20.004MiB; Debug Build) [SuccessX]
stefan@nuc ~/NEd $ 
```

### PR DESCRIPTION
**Pull request by cooldome**
I am not 100% confident on the approach. It is better if Araq reviews from design point of view.
So many TNoteKinds flying around, it has hard to say which one should take the priority.

### CODE REVIEW

**Comment by mratsim** [comment]
I suppose this also fixes #4044 (which was closed by workaround)

**Comment by Araq** [comment]
I'm thinking...

**Comment by mratsim** [comment]
New one related: #11852

File: compiler/pragmas.nim [hunk 1]
```diff
@@ -321,8 +321,12 @@ proc processNote(c: PContext, n: PNode) =
 
     let x = c.semConstBoolExpr(c, n[1])
     n.sons[1] = x
-    if x.kind == nkIntLit and x.intVal != 0: incl(c.config.notes, nk)
-    else: excl(c.config.notes, nk)
+    if x.kind == nkIntLit and x.intVal != 0:
+      incl(c.config.notes, nk)
```
Comment (Araq) [inline compiler/pragmas.nim]: This line seems to be the actual bugfix. The other code changes are a refactoring. The refactoring is ok, but is it required for the bugfix?

**Comment by cooldome** [comment]
I will describe the setup.
The problem comes from the fact that graph.config.notes is used to report messages and graph.config is reset with every import statement. AFAIK, it is done here https://github.com/nim-lang/Nim/blob/e6cb7e34effe7a47ab2a6d662e5bc99e89893757/compiler/modules.nim#L108.

Therefore, you loose your TNoteKinds on the first module import.

The fix is to use notes from TNoteKinds from PContext and make sure that whenever you set `graph.config.notes`, the `c.optionStack[^1].notes` is updated as well.

File: compiler/pragmas.nim (hunk 1 above)
Comment (cooldome) [reply compiler/pragmas.nim]: Other code changes are not refactoring. Please see my comment below.

**Comment by cooldome** [comment]
alternative fixes was submitted

**Pull request closed without merging**